_posts/2020-18-06-Triplestores-Onotologies-Reasoning.md

```diff
@@ -0,0 +1,354 @@
+this strategy, when each assertion, in this case each triple, is added
+to the Triplestore, the ontologies loaded will be consulted and any
+triples that would be created by the logical rules set by the ontology
+are created as are any other logical implications.  This means that
+all inferrencing is done at the time that the triple is added to the
+database.  The second is called _backwards-chaining_ which is
+implemented by Stardog.  In this strategy, the system uses a user
+query as a goal and works backwards from that goal.  If the goal cannot
+be logically inferred from the set of triples and the ontologies
+present in the Triplestore, nothing is returned.  Backwards-chaining
+is the inverse of forward-chaining.
+
+These two strategies have a large effect on how a user will perceive a
+system when querying it.  In the forward-chaining strategy, all
+inferences are done at the time that the files are added to the
+Triplestore.  This means that the repository on disk can be very large
+relative to the RDF files ingested.  For instance, at the time of
+writing, GraphDB's on disk representation of IrishGen, using
+forward-chaining, is 2.1 GB compared to the 4.4 MB size of the RDF
+files.  On the other hand, querying this database is very fast as all
+inferences are computed before a query is run.  Additionally,
+ingesting the RDF files can take a long time as the system will need
+to work out all the inferences, if any, for each triple.  In the
+backwards-chaining strategy, since the inferrencing is deferred until
+the time of the query, the on disk size is very small.  For instance
+in Stardog, the on disk representation of IrishGen is 81.3 MB.  The
+difference for users is felt in the speed of querying either of these
+Triplestores.  In Stardog, queries that may seem very straight forward
+can take an seemingly inordinate amount of time to complete.  However,
+this makes adding or deleting triples from the database very fast as
+the Stardog will not need to find and remove inferred statements.
+
+Which of these two systems the reader chooses to use is a matter of
+personal preference.  Both systems implement much of the standards,
+although each has its own details that the reader will need to
+understand while using it.  GraphDB is better if the user has the disk
+space to hold all the inferred triples and wishes to have very fast
+queries.  Stardog is better if the user wants more strict compliance
+with the standards and fast loading of IrishGen files, but queries can
+take a very long time to complete.
+
+## Conclusion
+
+The foregoing discussion has introduced the reader to Triplestores, a
+kind of graph database which is used to store and query RDF.  This led
+to a discussion of OWL 2 logical inferrencing and topics that a user
+who wants to write queries will need to know before attempting to use
+querying in Triplestores.  OWL 2 in conjunction with RDF gives a
+powerful set of tools with which to model the medieval Irish
```

Comment:
While this is true, for the model to produce any results, it requires a body of data encoded according to the model. So it might be worth pointing out that users can't just devise their own model and expect it to work with the RDF as it is at the moment.